source/optouts/OptOutsService.cs

```diff
@@ -0,0 +1,192 @@
+        {
+        }
+
+        /// <summary>
+        /// Gets a com.esendex.sdk.optouts.OptOut instance containing an opt out.
+        /// </summary>
+        /// <param name="id">A System.Guid instance that contains the Id of an opt out.</param>
+        /// <returns>A com.esendex.sdk.optouts.OptOut instance containing an opt out.</returns>
+        /// <exception cref="System.Net.WebException"></exception>
+        public OptOut GetById(Guid optOutId)
+        {
+            var requestUrl = new Uri(_baseUrl, string.Format("v1.0/optouts/{0}", optOutId));
+            var request = Request.Create("GET", requestUrl)
+                                 .WithHeader("Authorization", "Basic " + _credentials.EncodedValue())
+                                 .WithAcceptHeader(Constants.JSON_MEDIA_TYPE)
+                                 .If(_credentials.UseProxy, r => r.WithProxy(_credentials.WebProxy));
+
+            var response = request.GetResponse();
+            return response.DeserialiseJson<OptOut>();
+        }
+
+        /// <summary>
+        /// Gets a com.esendex.sdk.optouts.OptOutCollection instance containing a collection of com.esendex.sdk.optouts.OptOut.
+        /// </summary>
+        /// <param name="phoneNumber">A string that contains the phone number to search with.</param>
+        /// <param name="accountReference">A string that contains the account reference to search with.</param>
+        /// <param name="pageNumber">An int that that specifies which page of results to return.</param>
+        /// <param name="pageSize">An int that specifies how many results to return in a page.</param>
+        /// <returns>A com.esendex.sdk.optouts.OptOutCollection instance containing a collection of com.esendex.sdk.optouts.OptOut.</returns>
+        /// <exception cref="System.Net.WebException"></exception>
+        public OptOutCollection GetByPhoneNumber(string phoneNumber, string accountReference, int pageNumber, int pageSize)
+        {
+            var uri = BuildGetAllUri(phoneNumber, accountReference, pageNumber, pageSize);
+
+            var request = Request.Create("GET", uri)
+                                 .WithHeader("Authorization", "Basic " + _credentials.EncodedValue())
+                                 .WithAcceptHeader(Constants.JSON_MEDIA_TYPE)
+                                 .If(_credentials.UseProxy, r => r.WithProxy(_credentials.WebProxy));
+
+            var response = request.GetResponse();
+            return response.DeserialiseJson<OptOutCollection>();
+        }
+
+        private Uri BuildGetAllUri(string phoneNumber, string accountReference, int pageNumber, int pageSize)
+        {
+            var requestUrl = _baseUrl + "v1.0/optouts";
+
+            var builder = HttpUriBuilder.Create(requestUrl)
+                                        .WithParameter("startIndex", GetStartIndex(pageNumber, pageSize).ToString())

```

Comment:
Could this become something like:

```csharp
public class StartIndex : IParameter
    {
        private readonly int _pageNumber;
        private readonly int _pageSize;

        public StartIndex(int pageNumber, int pageSize)
        {
            _pageNumber = pageNumber;
            _pageSize = pageSize;
        }

        public Parameter CreateParameter()
        {
            if (_pageNumber < 1) throw new ArgumentException("Page number must be greater than zero.", "pageNumber");
                        if (_pageSize < 1) throw new ArgumentException("Page size must be greater than zero.", "pageSize");
            
            var startIndex = ((_pageNumber - 1) * _pageSize);
                     
            return new Parameter("startIndex", startIndex);
        }

    }
```

Then the WithParameter can invoke the create parameter internally and handle the values as such. This can be rolled out to all parameters?